---
title: "scRNAseq_wt"
author: "Masha"
date: "2024-06-13"
output: html_document
---
Here is the data set from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155543. They performed scRNAseq on 2 stages of *Drosophila melanogaster* larval development: 96h and 120h after egg layering (96H and 120H) with 2 replicates each. They used OregonR strain for that, which is the wild type strain, without any cassets inserted. For now, I am interested in 96H and will be analysing this time point first.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE)
```


```{r library setup}
#install.packages('Seurat')
library(Seurat)
library(here)
library(readr)
library(tidyverse)
library(org.Dm.eg.db)
#remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
library(DoubletFinder)
```

## Read data

```{r importing, include=FALSE}
d96h_h1 <-Read10X("../data/GSE155543_RAW/Dros96HH1/")
```


## Make a seurat object

```{r}
d96h_h1o<-CreateSeuratObject(counts = d96h_h1, project = "r1_96h", min.cells = 3, min.features = 200)
nrow(d96h_h1o)
```

```{r IDs-to-symbols}
gene_ids <-rownames(d96h_h1o)
#is.na(gene_ids) %>% sum() this to check whether any issues with rownames
gene_symbols<-  mapIds(org.Dm.eg.db, keys = Features(d96h_h1o), keytype = "FLYBASE", column="SYMBOL")

rownames(d96h_h1o) <- gene_symbols
```

## Estimating the ratio of mitochondrial genes.

```{r mito}
# outlook on the mitochondrial genes in the genome annotation. There are 87 of the corresponding gene names
mt_genes<-grepl("^m[A-Z]", gene_symbols)# %>% sum()
gene_symbols[mt_genes]

#Assigning the mitochondrial genes and calculating the percentage in the cells
d96h_h1o[["percent.mt"]] <- PercentageFeatureSet(d96h_h1o, pattern = "^m[A-Z]")

#Here nothing seems to be done? number of features and mitochondrial genes does not seem to fall far away from the median
VlnPlot(d96h_h1o, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r QC-mito}
#Here we see that the number of counts positively correlates with the number of mito genes. Thus, It is not clear to me where to set a threshold, definetely can be 0.9, but could it be lower?
plot1 <- FeatureScatter(d96h_h1o, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(d96h_h1o, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 
plot2
```
```{r}
#Here is very concervative filtering
d96h_h1o <- subset(d96h_h1o, subset = nFeature_RNA > 1000 & percent.mt < 0.9)
nrow(d96h_h1o)
```

## Normalising the data

```{r norm}
#Performing conventional normalisation
d96h_h1o <- NormalizeData(d96h_h1o, normalization.method = "LogNormalize", scale.factor = 10000)
```
## Identification of highly variable features

```{r}
# the default wa 2000 genes for human blood cells
d96h_h1o <- FindVariableFeatures(d96h_h1o, selection.method = "vst")

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(d96h_h1o), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(d96h_h1o)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1
plot2
```

## Scaling the data and dimentional reduction

```{r scaling}
d96h_h1o <- ScaleData(d96h_h1o)
```

```{r pca}
d96h_h1o <- RunPCA(d96h_h1o, features = VariableFeatures(object = d96h_h1o), npcs =30)
print(d96h_h1o[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(d96h_h1o, dims = 1:2, reduction = "pca")
DimPlot(d96h_h1o, reduction = "pca") + NoLegend()
```

```{r hm}
DimHeatmap(d96h_h1o, dims = 10:20, cells = 500, balanced = TRUE)
```

## Determine dimentionality of the dataset
```{r}
ElbowPlot(d96h_h1o)
```

```{r}
d96h_h1o <- FindNeighbors(d96h_h1o, dims = 1:12)
d96h_h1o <- FindClusters(d96h_h1o, resolution = 2)
```
```{r umap}
d96h_h1o <- RunUMAP(d96h_h1o, dims = 1:12)
DimPlot(d96h_h1o, reduction = "umap")
```

```{r}
FeaturePlot(d96h_h1o, features = c("SPARC", "Fas3", "ct", "zfh2", "nub", "vg", "brk", "dpp"))
```

Remove clusters with low quality cells
```{r}
# Extract the raw count matrix
raw_counts <- GetAssayData(d96h_h1o, assay = "RNA", slot = "counts")


genes_per_cell <- colSums(raw_counts > 0)
```

```{r}
# Create a dataframe with cluster identity and number of expressed genes per cell
cluster_gene_data <- data.frame(Cluster = Idents(d96h_h1o), GenesPerCell = genes_per_cell)

# Calculate the mean number of genes per cluster
cluster_gene_summary <- aggregate(GenesPerCell ~ Cluster, data = cluster_gene_data, FUN = mean)
sd_genes <- sd(genes_per_cell)
mean_genes<- mean(genes_per_cell)
threshold <- mean_genes-sd_genes
```

```{r}
low_gene_clusters <- subset(cluster_gene_summary, GenesPerCell < threshold)

d96_filtered <- subset(d96h_h1o, idents = low_gene_clusters$Cluster, invert = TRUE)
```

```{r}
d96_filtered <- RunUMAP(d96_filtered, dims = 1:12)
DimPlot(d96_filtered, reduction = "umap")
```
```{r}
whichCells<- function(data, x){
  temp<-WhichCells(data, idents = x)
  DimPlot(data, cells.highlight = temp)
}
#a vector with removed cluster names
active_clusters<-d96_filtered@active.ident %>% levels() %>% as.numeric()

# Apply the function for each cluster
plots <- lapply(active_clusters, function(x) whichCells(d96_filtered, x))

# Display all plots
print(plots)
#this is the cluster to be removed
print(plots[20])

#will give you the ids of all cells within a cluster which you can identify however you want
```

```{r}
#removing cluster n21 as the one containing cells in both groups
d96_filtered2 <- subset(d96_filtered, idents = 21, invert = TRUE)
d96_filtered2@active.ident %>% levels()
#how do I extract the cluster identity without counting the plots?
```

```{r}
d96_filtered2 <- RunUMAP(d96_filtered2, dims = 1:12)
DimPlot(d96_filtered2, reduction = "umap")
```

```{r}
active_clusters2<-d96_filtered2@active.ident %>% levels() %>% as.numeric()
plots <- lapply(active_clusters2, function(x) whichCells(d96_filtered2,x))
print(plots)
```

```{r}
#removing duplicates?

## pK Identification (no ground-truth) ----------------------------------------------------------------------------------
sweep.res.list <- paramSweep(d96_filtered, PCs = 1:12, sct = FALSE)
sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)
bcmvn <- find.pK(sweep.stats)
```


```{r}
#As I removed 4 clusters, do I need to re-run PCA? Do I need to re-mormalise?
FeaturePlot(d96_filtered2, features = c("SPARC", "Fas3", "wat", "tpr", "regucalcin", "He"))
```
```{r}
d96_filtered2 <- ScaleData(d96_filtered2)
d96_filtered2 <- RunPCA(d96_filtered2, features = VariableFeatures(object = d96_filtered2), npcs =30)
print(d96_filtered2[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(d96_filtered2, dims = 1:2, reduction = "pca")
DimPlot(d96_filtered2, reduction = "pca") + NoLegend()
```
```{r}
ElbowPlot(d96_filtered2)
```
```{r}
d96_filtered2 <- FindNeighbors(d96_filtered2, dims = 1:12)
d96_filtered2 <- FindClusters(d96_filtered2, resolution = 2)
```
```{r}
d96_filtered2 <- RunUMAP(d96_filtered2, dims = 1:12)
DimPlot(d96_filtered2, reduction = "umap")
```
```{r}
active_clusters3<-d96_filtered2@active.ident %>% levels() %>% as.numeric()
plots <- lapply(active_clusters3, function(x) whichCells(d96_filtered2,x))
print(plots)
```

```{r}
#re-assigning clusters to two big groups: epithelial and AMPs

new_ids<-c("AMP","AMP","AMP","AMP","epithelial","AMP","epithelial","AMP","epithelial","AMP", "epithelial","AMP", "AMP","AMP", "AMP", "epithelial","epithelial", "AMP", "AMP", "AMP", "epithelial","AMP","AMP", "epithelial","AMP","AMP")
length(new_ids)
names(new_ids)<-levels(d96_filtered2)
d96_filtered2 <- RenameIdents(d96_filtered2, new_ids)
DimPlot(d96_filtered2, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

```{r}
#do I need to recalculate all of those things?
d96_epithelial<- subset(d96_filtered2, idents ="epithelial")
d96_epithelial[["percent.mt"]] <- PercentageFeatureSet(d96_epithelial, pattern = "^m[A-Z]")

#Here nothing seems to be done? number of features and mitochondrial genes does not seem to fall far away from the median
VlnPlot(d96_epithelial, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
epithelial_genes<-FetchData(object = d96_epithelial, vars = "nFeature_RNA", layer = "counts")
epi_mean <- mean(epithelial_genes$nFeature_RNA)
epi_sd <- sd(epithelial_genes$nFeature_RNA)
d96_epi_filtered<-subset(d96_epithelial, subset=nFeature_RNA>(epi_mean-1.5*epi_sd)&nFeature_RNA<(epi_mean+1.5*epi_sd))
VlnPlot(d96_epi_filtered, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


```{r}
d96_epi_filtered <- ScaleData(d96_epi_filtered)
d96_epi_filtered <- RunPCA(d96_epi_filtered, features = VariableFeatures(object = d96_epi_filtered))
print(d96_epi_filtered[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(d96_epi_filtered, dims = 1:2, reduction = "pca")
DimPlot(d96_epi_filtered, reduction = "pca") + NoLegend()
ElbowPlot(d96_epi_filtered)
RunUMAP(d96_epi_filtered, dims = 1:12)
DimPlot(d96_epi_filtered, reduction = "umap")
```



```{r}
#extracting brk counts
brk<-FetchData(d96_epi_filtered, vars = "brk", layer="data") %>% as.data.frame()
#layer="counts" means raw data, layer="data" means normalised
hist(brk[,"brk"], 50)
```

