---
title: "d96H2"
author: "Masha"
date: "2025-02-07"
output: html_document
---

Here is the data set from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155543. They performed scRNAseq on 2 stages of *Drosophila melanogaster* larval development: 96h and 120h after egg layering (96H and 120H) with 2 replicates each. They used OregonR strain for that, which is the wild type strain, without any cassets inserted. For now, I am interested in 96H and will be analysing this time point first.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE)
```


```{r library setup}
#install.packages('Seurat')
library(Seurat)
library(here)
library(readr)
library(tidyverse)
library(org.Dm.eg.db)
library(loupeR)
library(scDblFinder)
```

## Read data

```{r importing, include=FALSE}
d96h_h2 <-Read10X("../data/GSE155543_RAW/Dros96H2/")
```

## Make a seurat object

```{r}
d96h_h2o<-CreateSeuratObject(counts = d96h_h2, project = "r2_96h", min.cells = 3, min.features = 200)
nrow(d96h_h2o) #number of genes, for cells there is a command Cells
```

```{r IDs-to-symbols}
gene_ids <-rownames(d96h_h2o)
#is.na(gene_ids) %>% sum() this to check whether any issues with rownames
gene_symbols<-  mapIds(org.Dm.eg.db, keys = Features(d96h_h2o), keytype = "FLYBASE", column="SYMBOL")

rownames(d96h_h2o) <- gene_symbols
head(rownames(d96h_h2o))
```

## Estimating the ratio of mitochondrial genes.

```{r mito}
# outlook on the mitochondrial genes in the genome annotation. There are 87 of the corresponding gene names
mt_genes<-grepl("^m[A-Z]", gene_symbols)# %>% sum()

#Assigning the mitochondrial genes and calculating the percentage in the cells
d96h_h2o[["percent.mt"]] <- PercentageFeatureSet(d96h_h2o, pattern = "^m[A-Z]")

#mitochondrial genes does not seem to fall far away from the median
VlnPlot(d96h_h2o, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


##  We want to check whether there are any doublets in our data set
Doublets are searched in unfiltered dataset, we use our original 10x matrix

```{r}
library(SingleCellExperiment)
suppressPackageStartupMessages(library(scDblFinder))
str(d96h_h2)
sce <- SingleCellExperiment(assays = list(counts = d96h_h2))
sce <- scDblFinder(sce)
sce$scDblFinder.score
sce$scDblFinder.class
str(sce)
```

### Now we are adding doublet classification to the metadata of unfiltered seurat object

```{r}
# Should return TRUE
all.equal(colnames(sce), colnames(d96h_h2o))
doublet_info <- sce$scDblFinder.class
d96h_h2o <- AddMetaData(d96h_h2o, metadata = doublet_info, col.name = "scDblFinder.class")
VlnPlot(d96h_h2o, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), group.by = "scDblFinder.class")
```


## Normalising the data

```{r norm}
#Performing conventional normalisation
anyDuplicated(rownames(d96h_h2o))
duplicates <- rownames(d96h_h2o)[duplicated(rownames(d96h_h2o))]
# print(duplicates)
 d96h_h2o <- d96h_h2o[!duplicated(rownames(d96h_h2o)), ]

d96h_h2o <- NormalizeData(d96h_h2o, normalization.method = "LogNormalize", scale.factor = 10000)

```
## Identification of highly variable features

```{r}
# the default wa 2000 genes for human blood cells
d96h_h2o <- FindVariableFeatures(d96h_h2o, selection.method = "vst")

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(d96h_h2o), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(d96h_h2o)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1
plot2
```

## Scaling the data and dimentional reduction

```{r scaling}
d96h_h2o <- ScaleData(d96h_h2o)
```

```{r pca}
d96h_h2o <- RunPCA(d96h_h2o, features = VariableFeatures(object = d96h_h2o), npcs =30)
print(d96h_h2o[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(d96h_h2o, dims = 1:2, reduction = "pca")
DimPlot(d96h_h2o, group.by="scDblFinder.class", reduction = "pca") + NoLegend()
```

```{r hm}
DimHeatmap(d96h_h2o, dims = 10:20, cells = 500, balanced = TRUE)
```

## Determine dimentionality of the dataset
```{r}
ElbowPlot(d96h_h2o)
```

```{r}
d96h_h2o <- FindNeighbors(d96h_h2o, dims = 1:16)
d96h_h2o <- FindClusters(d96h_h2o, resolution = 0.1)
```
```{r umap}
d96h_h2o <- RunUMAP(d96h_h2o, dims = 1:16)
DimPlot(d96h_h2o, reduction = "umap")
DimPlot(d96h_h2o, split.by="scDblFinder.class", reduction = "umap")
d96h_h2o[[]] %>% dplyr::count(seurat_clusters, scDblFinder.class) %>% spread(key='scDblFinder.class', value=n)
```
#from here it is not clear that there are doeblets. I need to filter the dataset in LoupeR and see what happens
```{r}
#library(loupeR)
#create_loupe_from_seurat(d96h_h2o, output_name = "../d96H2")
```


```{r}
#Here is filtering based on values from Rloupe
d96h_h2o <- subset(d96h_h2o, subset = nFeature_RNA > 2673 & nFeature_RNA < 4430 & nCount_RNA>19376 & nCount_RNA < 51383 & percent.mt < 0.8)
VlnPlot(d96h_h2o, group.by= "scDblFinder.class", features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

## Now as we made a few filtering steps, we need to re scale and re-run the analysis

```{r}
d96_filtered <- FindVariableFeatures(d96h_h2o, selection.method = "vst")
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(d96_filtered), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(d96_filtered)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)

plot2
```

```{r}
d96_filtered <- ScaleData(d96_filtered)
# Get features in scale.data
scaled_features <- rownames(GetAssayData(d96_filtered, slot = "scale.data"))

# Get features in new data layer
new_features <- rownames(GetAssayData(d96_filtered, slot = "data"))  # or "counts"

# Find missing features
missing_features <- setdiff(new_features, scaled_features)
extra_features <- setdiff(scaled_features, new_features)

print(missing_features)
print(extra_features)

```

```{r pca}
d96_filtered <- RunPCA(d96_filtered, features = VariableFeatures(object = d96_filtered), npcs =30)
print(d96_filtered[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(d96_filtered, dims = 1:2, reduction = "pca")
DimPlot(d96_filtered, group.by = "scDblFinder.class", reduction = "pca") + NoLegend()
```
### we see much better separation and few predicted doublets. We will keep them, because it is just a prediction and we already done some harsh filtering

```{r}
ElbowPlot(d96_filtered)
```
```{r}
d96_filtered <- FindNeighbors(d96_filtered, dims = 1:13)
d96_filtered <- FindClusters(d96_filtered, resolution = 0.1)
```

```{r}
d96_filtered <- RunUMAP(d96_filtered, dims = 1:13)
DimPlot(d96_filtered, reduction = "umap")
DimPlot(d96_filtered, split.by="scDblFinder.class", reduction = "umap")
```
#### As I saw doublets in-between two clusters I want to remove them and see how it affects PCA

```{r}
d96_filtered_doublets <- subset(d96_filtered, subset=scDblFinder.class=="singlet")
d96_filtered_doublets <- FindVariableFeatures(d96_filtered_doublets, selection.method = "vst") %>% ScaleData()
d96_filtered_doublets <- RunPCA(d96_filtered_doublets, features = VariableFeatures(object = d96_filtered_doublets), npcs =30)
print(d96_filtered_doublets[["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(d96_filtered_doublets, dims = 1:2, reduction = "pca")
DimPlot(d96_filtered_doublets, group.by = "scDblFinder.class", reduction = "pca") + NoLegend()
```

```{r}
d96_filtered_doublets <- FindNeighbors(d96_filtered_doublets, dims = 1:13)
d96_filtered_doublets <- FindClusters(d96_filtered_doublets, resolution = 0.1)
```
```{r}
d96_filtered_doublets <- RunUMAP(d96_filtered_doublets, dims = 1:13)
DimPlot(d96_filtered_doublets, reduction = "umap")
```

```{r}
 FeaturePlot(d96_filtered_doublets, features = c("SPARC", "Fas3", "eyg", "zfh2", "nub"), order=TRUE)
#d96_filtered[[]] %>% dplyr::count(seurat_clusters, scDblFinder.class) %>% spread(key='scDblFinder.class', value=n)
```
### Saving filtered seurat
------------------------------------------------------------here I stopped
```{r}
SaveSeuratRds(d96_filtered_doublets, "../data/d96H2_filtered.Rds")
```


### We need to work only with epithelial cells. We filter them

```{r}
#subsetting epithelial cells cluster
d96_epithelial <- subset(d96_filtered_doublets, idents = 2)

d96_epithelial <- FindVariableFeatures(d96_epithelial, selection.method = "vst")
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(d96_epithelial), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(d96_epithelial)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE, ynudge=0, xnudge=0)
plot2
```

```{r}
d96_epithelial <- ScaleData(d96_epithelial)
d96_epithelial  <- RunPCA(d96_epithelial , features = VariableFeatures(object = d96_epithelial ), npcs =30)
print(d96_epithelial [["pca"]], dims = 1:5, nfeatures = 5)

VizDimLoadings(d96_epithelial , dims = 1:2, reduction = "pca")
DimPlot(d96_epithelial , group.by = "scDblFinder.class", reduction = "pca") + NoLegend()
```

```{r}
ElbowPlot(d96_epithelial)
```

```{r}
d96_epithelial <- FindNeighbors(d96_epithelial, dims = 1:16) %>%
  FindClusters(resolution = 0.3)
```
```{r}
d96_epithelial <- RunUMAP(d96_epithelial, dims = 1:16)
DimPlot(d96_epithelial, reduction = "umap")
```
```{r}
FeaturePlot(d96_epithelial, features = c("nub", "rn"), order=TRUE) #pouch
```
```{r}
FeaturePlot(d96_epithelial, features = c("eyg", "pnr"), order=TRUE) #notum
```
modulescores-> takes a few genes and plots avg value against random genes
Seurat::FindSubCluster()
number of dimentions and cluster resolution?

glmgampoi for brk diff expression

```{r}
FeaturePlot(d96_epithelial, features = c("zfh2", "Sox15"), order=TRUE) #hinge
```


```{r}
FeaturePlot(d96_epithelial, features = c("Ubx", "eya"), order=TRUE) #PE
```


```{r}
markers<- FindAllMarkers(d96_epithelial, only.pos = TRUE)
markers0.05 <- markers %>% 
  filter(p_val_adj<0.05)

top_markers <- markers0.05 %>% 
  group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC, n=12)

DotPlot(d96_epithelial, features = top_markers$gene) + RotatedAxis() #you can do that manually just add features you want from the paper
```


```{r}
levels(d96_epithelial)
new_ident <- c("pouch", "hinge", "notum", "PE")
names(new_ident) <- levels(d96_epithelial)
d96_epithelial<- RenameIdents(d96_epithelial, new_ident)
DimPlot(d96_epithelial, reduction = "umap")
```

###for further analysis of Brk diff expr we need to remove PE
```{r}
d96_dp <- subset(d96_epithelial, idents= "PE", invert=TRUE)
DimPlot(d96_dp, reduction="umap")
```


```{r}
VlnPlot(d96_dp, features = c("brk", "bi"))
```

```{r}
d96_pouch <-subset(d96_epithelial, idents="pouch")
d96_pouch <- ScaleData(d96_pouch) %>% 
  RunPCA(features = VariableFeatures(object = d96_pouch), npcs =30)

VizDimLoadings(d96_pouch , dims = 1:2, reduction = "pca")
DimPlot(d96_pouch, reduction = "pca") + NoLegend()
#its rather a spectrum
```

```{r}
ElbowPlot(d96_pouch)
```

```{r}
d96_pouch <- FindNeighbors(d96_pouch, dims = 1:16) %>%
  FindClusters(resolution = 0.3)
```
```{r}
d96_pouch <- RunUMAP(d96_pouch, dims = 1:16)
DimPlot(d96_pouch, reduction = "umap")
```

```{r}
markers_pouch<- FindAllMarkers(d96_pouch, only.pos = TRUE)
markers_pouch0.05 <- markers_pouch %>% 
  filter(p_val_adj<0.05)

top_markers_pouch <- markers_pouch0.05 %>% 
  group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC, n=10)

DotPlot(d96_pouch, features = top_markers_pouch$gene) + RotatedAxis()
```


```{r}
FeaturePlot(d96_pouch, c("bi", "brk"), order=TRUE)
```

```{r}
brk_pouch <- FetchData(d96_pouch, vars = "brk", layer="data") %>% as.data.frame()
#layer="counts" means raw data, layer="data" means normalised

#plot(density(brk_pouch$brk))
```


```{r}
dens_brk<-density(brk_pouch$brk)
y_c<-dens_brk$y
peak.indices <- which(diff(sign(diff(y_c))) == -2) + 1
# Extract the peak heights and corresponding x positions
peak.heights <- y_c[peak.indices]
peak.positions <- dens_brk$x[peak.indices]

ordered.peaks <- order(peak.heights, decreasing = TRUE)
secondary.peak <- list(x = peak.positions[ordered.peaks[2]], 
                       y = peak.heights[ordered.peaks[2]])
```

```{r}
plot(dens_brk, main = "Density Estimate with Peaks")
points(secondary.peak$x, secondary.peak$y, col = "blue", pch = 19, cex = 1.5)
cat("Secondary Peak: x =", secondary.peak$x, "y =", secondary.peak$y, "\n")
```

```{r}
d96_pouch$Brk_expression <- ifelse(brk_pouch$brk > 0.2, "High", "Low")
DimPlot(d96_pouch, group.by = "Brk_expression")
```

```{r}
FeaturePlot(d96_pouch, c("brk"), order=TRUE, split.by = "Brk_expression")
#Dotplot(d96_pouch, )
```

```{r}
pouch_brk_markers <- FindMarkers(d96_pouch, ident.1 = "High", ident.2 = "Low", group.by = "Brk_expression")
pouch_brk_markers_filtered<- pouch_brk_markers %>% 
  filter(p_val_adj<0.05) %>% 
  arrange(avg_log2FC)
```

```{r}
pouch_brk_markers %>% 
  rownames_to_column("gene") %>% 
  arrange(desc(avg_log2FC)) %>% 
  na.omit() %>% 
  dplyr::select(gene, avg_log2FC) %>% 
  deframe()-> pouch_brk_markers_rank
```

```{r}
library(clusterProfiler)
library(enrichplot)

gse <- gseGO(pouch_brk_markers_rank,
             ont="BP",
             OrgDb=org.Dm.eg.db,
             keyType="SYMBOL",
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             minGSSize=3,
             pAdjustMethod = "BH",
             nPermSimple = 10000)
summary(gse) #%>% as.tibble() -> gse_summary

```

### Bi analysis
```{r}
bi_pouch <- FetchData(d96_pouch, vars = "bi", layer="data") %>% as.data.frame()
#layer="counts" means raw data, layer="data" means normalised
plot(density(bi_pouch$bi))
```

```{r}
d96_pouch$bi_expression <- ifelse(bi_pouch$bi > 1, "High", "Low")
DimPlot(d96_pouch, group.by = "bi_expression")

```

```{r}
FeaturePlot(d96_pouch, "bi", order=TRUE, split.by = "bi_expression")
```


```{r}
pouch_bi_markers <- FindMarkers(d96_pouch, ident.1 = "High", ident.2 = "Low", group.by = "bi_expression")
pouch_bi_markers_filtered<- pouch_bi_markers %>% 
  filter(p_val_adj<0.05) %>% 
  arrange(avg_log2FC)
```

```{r}
pouch_bi_markers %>% 
  rownames_to_column("gene") %>% 
  arrange(desc(avg_log2FC)) %>% 
  na.omit() %>% 
  dplyr::select(gene, avg_log2FC) %>% 
  deframe()-> pouch_bi_markers_rank
```

```{r}
gse <- gseGO(pouch_bi_markers_rank,
             ont="BP",
             OrgDb=org.Dm.eg.db,
             keyType="SYMBOL",
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             minGSSize=3,
             pAdjustMethod = "BH")
as.data.frame(gse) %>% as.tibble() -> gse_summary
```
```{r}
export_markers<-pouch_brk_markers_filtered %>% mutate(., gene = mapIds(
  org.Dm.eg.db,
  keys = rownames(.),
  column ="FLYBASE",
  keytype = "SYMBOL",
  multiVals = "first")) %>% rownames_to_column()
  
write_csv(export_markers, file="../pouch_brk_markersh2.csv") 
```

```{r}
saveRDS(d96_pouch, file = "../data/d96_pouch_h2.Rds")
```

